### PR TITLE
Batched generation

### PR DESCRIPTION
**Pull request by lvwerra**
This enables to generate in batches rather than one by one with a custom batch size. This simplifies the generation and makes generation significantly faster! Changes are backwards compatible.

Code before:
```python
response_tensors = []
for question in tqdm(question_tensors):
    gen_len = output_length_sampler()
    generation_kwargs["max_new_tokens"] = gen_len
    response = ppo_trainer.generate(question, **generation_kwargs)
    response_tensors.append(response.squeeze()[-gen_len:])
batch["response"] = [tokenizer.decode(r.squeeze()) for r in response_tensors]
```

Code after:
```python
response_tensors = ppo_trainer.generate(question_tensors,  return_prompt=False,
    length_sampler= output_length_sampler,  **generation_kwargs)
batch["response"] = tokenizer.batch_decode(response_tensors)
```

**Todo:** Add some tests.

### CODE REVIEW

**Comment by lvwerra** [comment]
Added some test and ran the following benchmark (cc @edbeeching @natolambert) on the generation part:

| Model    | New tokens | Batch size | Samples | Time   |
|----------|------------|------------|---------|--------|
| GPT-2    | 64         | 1          | 256     | 176.08 |
| GPT-2    | 64         | 2          | 256     | 91.54  |
| GPT-2    | 64         | 4          | 256     | 37.29  |
| GPT-2    | 64         | 8          | 256     | 23.93  |
| GPT-2    | 64         | 16         | 256     | 11.58  |
|          |            |            |         |        |
| GPT-2 XL | 64         | 1          | 256     | 568.64 |
| GPT-2 XL | 64         | 4          | 256     | 379.76 |
| GPT-2 XL | 64         | 4          | 256     | 234.63 |
| GPT-2 XL | 64         | 8          | 256     | 176.02 |

**Review by younesbelkada — COMMENTED** [review]
Awesome speedup! Looks great in general 🚀 
I left some comments, I think we need a safety checker to check if `batch_size` is effectively smaller than `len(queries)`

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -371,11 +378,67 @@ def generate(self, query_tensor: torch.Tensor, **generation_kwargs):
         Returns:
             `torch.LongTensor`: A tensor of shape (`batch_size`, `gen_len`) containing response tokens.
         """
-        response = self.accelerator.unwrap_model(self.model).generate(
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:374]: good point!

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -371,11 +378,67 @@ def generate(self, query_tensor: torch.Tensor, **generation_kwargs):
...
+        return_prompt: bool = True,
+        pad_to_multiple_of: int = None,
+        **generation_kwargs,
+    ):
+        outputs = []
+
+        padding_side_default = self.tokenizer.padding_side
+        self.tokenizer.padding_side = "left"
+
+        for i in range(0, len(query_tensors), batch_size):
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:416]: yep! `batch_size = min(len(query_tensors), batch_size)`?

File: trl/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -371,11 +378,67 @@ def generate(self, query_tensor: torch.Tensor, **generation_kwargs):
...
+                pad_to_multiple_of=pad_to_multiple_of,
+                return_tensors="pt",
+            ).to(self.current_device)
+
+            generations = self.accelerator.unwrap_model(self.model).generate(**padded_inputs, **generation_kwargs)
+
+            for generation, mask in zip(generations, padded_inputs["attention_mask"]):
+                output = generation[(1 - mask).sum() :]  # remove padding
+                if not return_prompt:
+                    output = output[(mask).sum() :]  # remove prompt
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:437]: we already have the query and it all examples we spend 1-2 lines removing the queries from the generations. with that it's done automatically :)

File: trl/trainer/ppo_trainer.py [hunk 4]
```diff
@@ -371,11 +378,67 @@ def generate(self, query_tensor: torch.Tensor, **generation_kwargs):
...
+        padding_side_default = self.tokenizer.padding_side
+        self.tokenizer.padding_side = "left"
+
+        for i in range(0, len(query_tensors), batch_size):
+            if length_sampler is not None:
+                generation_kwargs["max_new_tokens"] = length_sampler()
+
+            batch = query_tensors[i : i + batch_size]
+            batch_mask = [torch.ones_like(element) for element in batch]
+            inputs = {"input_ids": batch, "attention_mask": batch_mask}
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:422]: what do you mean? the `pad` extends the attention mask with zeros for the padding. if we don't pass it we don't have one at all.

File: trl/trainer/ppo_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py:437]: AHh good point ok, you mean [here](https://github.com/lvwerra/trl/blob/0610711ddab3ba1d8b5d41d31423c213b433472e/examples/sentiment/scripts/gpt2-sentiment.py#L177) right?
In this case for the API consistency I think that we need to add it as well in the block that does not called batched generate as well.
Also note that for seq2seq models, the model already returns the response without the query: https://github.com/lvwerra/trl/blob/0610711ddab3ba1d8b5d41d31423c213b433472e/examples/sentiment/scripts/t5-sentiment.py#L153 so might be worth it to add another safety checker `if self.is_encoder_decoder`

[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:437]: > In this case for the API consistency I think that we need to add it as well in the block that does not call batched generate as well.

We do here, no?: https://github.com/lvwerra/trl/blob/ee04bada9d4607c8273b41119d9adde98c7c9528/trl/trainer/ppo_trainer.py#L398

Good point about enc-dec, will add an extra clause.

[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py:437]: Ah yes thanks, indeed we do it

**Review by younesbelkada — APPROVED** [review]
Awesome work! Thanks a lot for this!
